### ISSUE TITLE
/sync panics after joining a room

### ISSUE THREAD

**Issue by Half-Shot** [matrix-org/dendrite#1015]
I joined `!XKjkxFMpKwcGmVfsvN:cadair.com`, however ever since the sync has been panicing. The panic line is:

```
time="2020-05-08T15:17:41.824953851Z" level=error msg="Request panicked!
goroutine 316 [running]:
runtime/debug.Stack(0xc0001040e0, 0xc0005e1dc8, 0xc0001041c0)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
github.com/matrix-org/util.Protect.func1.1(0xc000130500, 0x1098440, 0xc0001442a0)
	/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:98 +0x15d
panic(0xe85ce0, 0xc0000360e0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/matrix-org/dendrite/syncapi/storage/postgres.(*SyncServerDatasource).getResponseWithPDUsForCompleteSync(0xc0001a5680, 0x109c340, 0xc0003d13b0, 0xc00064e3a0, 0x1b, 0x14, 0xc000644240, 0x0, 0x0, 0x23e, ...)
	/build/syncapi/storage/postgres/syncserver.go:590 +0xbf5
github.com/matrix-org/dendrite/syncapi/storage/postgres.(*SyncServerDatasource).CompleteSync(0xc0001a5680, 0x109c340, 0xc0003d13b0, 0xc00064e3a0, 0x1b, 0x14, 0xf078af, 0x7, 0xc0001a49e8)
	/build/syncapi/storage/postgres/syncserver.go:622 +0x81
github.com/matrix-org/dendrite/syncapi/sync.(*RequestPool).currentSyncForUser(0xc000185410, 0x109c340, 0xc0003d13b0, 0xc00036c2b8, 0x8, 0xc00064e3a0, 0x1b, 0xc0003f8407, 0x2b, 0x8, ...)
	/build/syncapi/sync/requestpool.go:138 +0x376
github.com/matrix-org/dendrite/syncapi/sync.(*RequestPool).OnIncomingSyncRequest(0xc000185410, 0xc000130700, 0xc0000c82d0, 0x0, 0x0, 0x0, 0x0)
	/build/syncapi/sync/requestpool.go:67 +0x40c
github.com/matrix-org/dendrite/syncapi/routing.Setup.func1(0xc000130700, 0xc0000c82d0, 0xd8a100, 0x106f3d0, 0xef7f20, 0xc00018c850)
	/build/syncapi/routing/routing.go:56 +0x3e
github.com/matrix-org/dendrite/common.MakeAuthAPI.func1(0xc000130600, 0x0, 0x0, 0x0, 0x0)
	/build/common/httpapi.go:46 +0x325
github.com/matrix-org/util.(*jsonRequestHandlerWrapper).OnIncomingRequest(0xc000010090, 0xc000130600, 0xc0003d0d20, 0x30, 0x30, 0xe7a7c0)
	/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:79 +0x33
github.com/matrix-org/util.MakeJSONAPI.func1(0x1098440, 0xc0001442a0, 0xc000130600)
	/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:139 +0x97
github.com/matrix-org/util.Protect.func1(0x1098440, 0xc0001442a0, 0xc000130500)
	/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:103 +0xa1
net/http.HandlerFunc.ServeHTTP(...)
	/usr/local/go/src/net/http/server.go:2007
github.com/matrix-org/dendrite/common.MakeExternalAPI.func1(0x1098440, 0xc0001442a0, 0xc000130400)
	/build/common/httpapi.go:106 +0x22b
net/http.HandlerFunc.ServeHTTP(0xc000230240, 0x1098440, 0xc0001442a0, 0xc000130400)
	/usr/local/go/src/net/http/server.go:2007 +0x44
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00018a240, 0x1098440, 0xc0001442a0, 0xc000130100)
	/go/pkg/mod/github.com/gorilla/mux@v1.7.3/mux.go:212 +0xe2
github.com/matrix-org/dendrite/common.WrapHandlerInCORS.func1(0x1098440, 0xc0001442a0, 0xc000130100)
	/build/common/httpapi.go:231 +0x340
net/http.HandlerFunc.ServeHTTP(0xc00024ea60, 0x1098440, 0xc0001442a0, 0xc000130100)
	/usr/local/go/src/net/http/server.go:2007 +0x44
net/http.(*ServeMux).ServeHTTP(0x16b1620, 0x1098440, 0xc0001442a0, 0xc000130100)
	/usr/local/go/src/net/http/server.go:2387 +0x1bd
net/http.serverHandler.ServeHTTP(0xc000148000, 0x1098440, 0xc0001442a0, 0xc000130100)
	/usr/local/go/src/net/http/server.go:2802 +0xa4
net/http.(*conn).serve(0xc000273720, 0x109c280, 0xc0005da140)
	/usr/local/go/src/net/http/server.go:1890 +0x875
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2928 +0x384
" func="1
	" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:97]" context=missing panic="runtime error: index out of range [0] with length 0"
```

I have verified that this is the room causing the panic

**Comment by Half-Shot** [comment]
Update: So after talking in this room from another homeserver, the sync started working.

```
time="2020-05-08T16:13:36.539092427Z" level=info msg="Processed 0 PDUs from transaction \"1588804030455\"" func="processTransaction\n\t" file=" [/build/federationapi/routing/send.go:186]" req.id=xzBdsy9QNN9U req.method=PUT req.path=/_matrix/federation/v1/send/1588804030455
time="2020-05-08T16:13:36.539133140Z" level=info msg="Responding (11 bytes)" func="respond\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:170]" code=200 req.id=xzBdsy9QNN9U req.method=PUT req.path=/_matrix/federation/v1/send/1588804030455
time="2020-05-08T16:13:37.342250539Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:122]" req.id=dF5nhIb9iGXB req.method=GET req.path=/_matrix/client/versions
time="2020-05-08T16:13:37.342351710Z" level=info msg="Responding (50 bytes)" func="respond\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:170]" code=200 req.id=dF5nhIb9iGXB req.method=GET req.path=/_matrix/client/versions
time="2020-05-08T16:13:37.414198804Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:122]" req.id=X5w45dJnoGru req.method=GET req.path=/_matrix/client/r0/sync
Problem room: !XKjkxFMpKwcGmVfsvN:cadair.com
time="2020-05-08T16:13:37.462565240Z" level=error msg="Request panicked!\ngoroutine 1104 [running]:\nruntime/debug.Stack(0xc0001222a0, 0xc0004c5db8, 0xc000122310)\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x9d\ngithub.com/matrix-org/util.Protect.func1.1(0xc000130700, 0x1098680, 0xc00068f7a0)\n\t/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:98 +0x15d\npanic(0xe85ce0, 0xc000289a80)\n\t/usr/local/go/src/runtime/panic.go:679 +0x1b2\ngithub.com/matrix-org/dendrite/syncapi/storage/postgres.(*SyncServerDatasource).getResponseWithPDUsForCompleteSync(0xc0000d3d40, 0x109c580, 0xc0003dda70, 0xc00038bd00, 0x1b, 0x14, 0xc00015ec00, 0x0, 0x0, 0x244, ...)\n\t/build/syncapi/storage/postgres/syncserver.go:592 +0xc9a\ngithub.com/matrix-org/dendrite/syncapi/storage/postgres.(*SyncServerDatasource).CompleteSync(0xc0000d3d40, 0x109c580, 0xc0003dda70, 0xc00038bd00, 0x1b, 0x14, 0xf078af, 0x7, 0xc0007320e8)\n\t/build/syncapi/storage/postgres/syncserver.go:624 +0x81\ngithub.com/matrix-org/dendrite/syncapi/sync.(*RequestPool).currentSyncForUser(0xc000134e40, 0x109c580, 0xc0003dda70, 0xc000370538, 0x8, 0xc00038bd00, 0x1b, 0xc0004286c7, 0x2b, 0xc, ...)\n\t/build/syncapi/sync/requestpool.go:138 +0x376\ngithub.com/matrix-org/dendrite/syncapi/sync.(*RequestPool).OnIncomingSyncRequest(0xc000134e40, 0xc000130900, 0xc00043cd20, 0x0, 0x0, 0x0, 0x0)\n\t/build/syncapi/sync/requestpool.go:67 +0x40c\ngithub.com/matrix-org/dendrite/syncapi/routing.Setup.func1(0xc000130900, 0xc00043cd20, 0xd8a100, 0x106f610, 0xef7f20, 0xc000122690)\n\t/build/syncapi/routing/routing.go:56 +0x3e\ngithub.com/matrix-org/dendrite/common.MakeAuthAPI.func1(0xc000130800, 0x0, 0x0, 0x0, 0x0)\n\t/build/common/httpapi.go:46 +0x325\ngithub.com/matrix-org/util.(*jsonRequestHandlerWrapper).OnIncomingRequest(0xc0000c4028, 0xc000130800, 0xc0003dd050, 0x30, 0x30, 0xe7a7c0)\n\t/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:79 +0x33\ngithub.com/matrix-org/util.MakeJSONAPI.func1(0x1098680, 0xc00068f7a0, 0xc000130800)\n\t/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:139 +0x97\ngithub.com/matrix-org/util.Protect.func1(0x1098680, 0xc00068f7a0, 0xc000130700)\n\t/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:103 +0xa1\nnet/http.HandlerFunc.ServeHTTP(...)\n\t/usr/local/go/src/net/http/server.go:2007\ngithub.com/matrix-org/dendrite/common.MakeExternalAPI.func1(0x1098680, 0xc00068f7a0, 0xc000130600)\n\t/build/common/httpapi.go:106 +0x22b\nnet/http.HandlerFunc.ServeHTTP(0xc0000adbf0, 0x1098680, 0xc00068f7a0, 0xc000130600)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/gorilla/mux.(*Router).ServeHTTP(0xc0001cc180, 0x1098680, 0xc00068f7a0, 0xc000130100)\n\t/go/pkg/mod/github.com/gorilla/mux@v1.7.3/mux.go:212 +0xe2\ngithub.com/matrix-org/dendrite/common.WrapHandlerInCORS.func1(0x1098680, 0xc00068f7a0, 0xc000130100)\n\t/build/common/httpapi.go:231 +0x340\nnet/http.HandlerFunc.ServeHTTP(0xc00021e480, 0x1098680, 0xc00068f7a0, 0xc000130100)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\nnet/http.(*ServeMux).ServeHTTP(0x16b2620, 0x1098680, 0xc00068f7a0, 0xc000130100)\n\t/usr/local/go/src/net/http/server.go:2387 +0x1bd\nnet/http.serverHandler.ServeHTTP(0xc000162000, 0x1098680, 0xc00068f7a0, 0xc000130100)\n\t/usr/local/go/src/net/http/server.go:2802 +0xa4\nnet/http.(*conn).serve(0xc00024d720, 0x109c4c0, 0xc0006d19c0)\n\t/usr/local/go/src/net/http/server.go:1890 +0x875\ncreated by net/http.(*Server).Serve\n\t/usr/local/go/src/net/http/server.go:2928 +0x384\n" func="1\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:97]" context=missing panic="runtime error: index out of range [0] with length 0"
time="2020-05-08T16:13:37.462663154Z" level=info msg="Responding (35 bytes)" func="respond\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:170]" code=500 context=missing
time="2020-05-08T16:13:39.505975365Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:122]" req.id=Tmbnlw38uf4J req.method=PUT req.path=/_matrix/federation/v1/send/1588095882272
time="2020-05-08T16:13:39.506545899Z" level=info msg="Received transaction \"1588095882272\" containing 0 PDUs, 1 EDUs" func="Send\n\t" file=" [/build/federationapi/routing/send.go:72]" req.id=Tmbnlw38uf4J req.method=PUT req.path=/_matrix/federation/v1/send/1588095882272
time="2020-05-08T16:13:39.506570258Z" level=warning msg="unhandled edu" func="processEDUs\n\t" file=" [/build/federationapi/routing/send.go:225]" req.id=Tmbnlw38uf4J req.method=PUT req.path=/_matrix/federation/v1/send/1588095882272 type=m.receipt
time="2020-05-08T16:13:39.506585020Z" level=info msg="Processed 0 PDUs from transaction \"1588095882272\"" func="processTransaction\n\t" file=" [/build/federationapi/routing/send.go:186]" req.id=Tmbnlw38uf4J req.method=PUT req.path=/_matrix/federation/v1/send/1588095882272
time="2020-05-08T16:13:39.506604809Z" level=info msg="Responding (11 bytes)" func="respond\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:170]" code=200 req.id=Tmbnlw38uf4J req.method=PUT req.path=/_matrix/federation/v1/send/1588095882272
time="2020-05-08T16:13:42.629813702Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:122]" req.id=3FC3NmEkK3F7 req.method=PUT req.path=/_matrix/federation/v1/send/1588095882282
time="2020-05-08T16:13:42.630455028Z" level=info msg="Received transaction \"1588095882282\" containing 1 PDUs, 0 EDUs" func="Send\n\t" file=" [/build/federationapi/routing/send.go:72]" req.id=3FC3NmEkK3F7 req.method=PUT req.path=/_matrix/federation/v1/send/1588095882282
time="2020-05-08T16:13:42.651551690Z" level=info msg="received event from roomserver" func="onNewRoomEvent\n\t" file=" [/build/syncapi/consumers/roomserver.go:106]" event_id="$PA9TaLar4KH7XAPuePfOriJP-hxkF0b4V26uBU9nBzM" room_id="!XKjkxFMpKwcGmVfsvN:cadair.com" room_version=5
time="2020-05-08T16:13:42.651794636Z" level=info msg="received room event from roomserver" func="onMessage\n\t" file=" [/build/federationsender/consumers/roomserver.go:93]" event_id="$PA9TaLar4KH7XAPuePfOriJP-hxkF0b4V26uBU9nBzM" room_id="!XKjkxFMpKwcGmVfsvN:cadair.com" send_as_server=""
time="2020-05-08T16:13:42.651838490Z" level=info msg="received event from roomserver" func="onMessage\n\t" file=" [/build/clientapi/consumers/roomserver.go:92]" event_id="$PA9TaLar4KH7XAPuePfOriJP-hxkF0b4V26uBU9nBzM" room_id="!XKjkxFMpKwcGmVfsvN:cadair.com" type=m.room.message
time="2020-05-08T16:13:42.651838284Z" level=info msg="received event from roomserver" func="onMessage\n\t" file=" [/build/publicroomsapi/consumers/roomserver.go:86]" event_id="$PA9TaLar4KH7XAPuePfOriJP-hxkF0b4V26uBU9nBzM" room_id="!XKjkxFMpKwcGmVfsvN:cadair.com" type=m.room.message
time="2020-05-08T16:13:42.651799375Z" level=info msg="appservice received an event from roomserver" func="onMessage\n\t" file=" [/build/appservice/consumers/roomserver.go:99]" event_id="$PA9TaLar4KH7XAPuePfOriJP-hxkF0b4V26uBU9nBzM" room_id="!XKjkxFMpKwcGmVfsvN:cadair.com" type=m.room.message
time="2020-05-08T16:13:42.652351161Z" level=info msg="Processed 1 PDUs from transaction \"1588095882282\"" func="processTransaction\n\t" file=" [/build/federationapi/routing/send.go:186]" req.id=3FC3NmEkK3F7 req.method=PUT req.path=/_matrix/federation/v1/send/1588095882282
time="2020-05-08T16:13:42.652403710Z" level=info msg="Responding (60 bytes)" func="respond\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:170]" code=200 req.id=3FC3NmEkK3F7 req.method=PUT req.path=/_matrix/federation/v1/send/1588095882282
time="2020-05-08T16:13:43.375941975Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:122]" req.id=gOKaf0PUzO03 req.method=PUT req.path=/_matrix/federation/v1/send/1588819156560
time="2020-05-08T16:13:43.377854155Z" level=info msg="Received transaction \"1588819156560\" containing 0 PDUs, 1 EDUs" func="Send\n\t" file=" [/build/federationapi/routing/send.go:72]" req.id=gOKaf0PUzO03 req.method=PUT req.path=/_matrix/federation/v1/send/1588819156560
time="2020-05-08T16:13:43.377938158Z" level=warning msg="unhandled edu" func="processEDUs\n\t" file=" [/build/federationapi/routing/send.go:225]" req.id=gOKaf0PUzO03 req.method=PUT req.path=/_matrix/federation/v1/send/1588819156560 type=m.presence
time="2020-05-08T16:13:43.377988298Z" level=info msg="Processed 0 PDUs from transaction \"1588819156560\"" func="processTransaction\n\t" file=" [/build/federationapi/routing/send.go:186]" req.id=gOKaf0PUzO03 req.method=PUT req.path=/_matrix/federation/v1/send/1588819156560
time="2020-05-08T16:13:43.378070143Z" level=info msg="Responding (11 bytes)" func="respond\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:170]" code=200 req.id=gOKaf0PUzO03 req.method=PUT req.path=/_matrix/federation/v1/send/1588819156560
time="2020-05-08T16:13:43.491032859Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:122]" req.id=fIz2mWCssPrE req.method=GET req.path=/_matrix/client/versions
time="2020-05-08T16:13:43.491141874Z" level=info msg="Responding (50 bytes)" func="respond\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:170]" code=200 req.id=fIz2mWCssPrE req.method=GET req.path=/_matrix/client/versions
time="2020-05-08T16:13:43.535085776Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:122]" req.id=bZmE0yvPqGRm req.method=GET req.path=/_matrix/client/r0/sync
Problem room: !lTYKYBkCnN08Bzp8:den.half-shot.uk
Problem room: !uXDCzlYgCTHtiWCkEx:jki.re
Problem room: !RlJWMfbSUlPhEubpsc:half-shot.uk
Problem room: !VJwUGwLFWvNmfOvyxL:reyd.us
Problem room: !XKjkxFMpKwcGmVfsvN:cadair.com
time="2020-05-08T16:13:43.589295146Z" level=info msg="Responding immediately" func="OnIncomingSyncRequest\n\t" file=" [/build/syncapi/sync/requestpool.go:72]" next=s581_0 req.id=bZmE0yvPqGRm req.method=GET req.path=/_matrix/client/r0/sync since="<nil>" timeout=0s userID="@half-shot:den.half-shot.uk" user_id="@half-shot:den.half-shot.uk"
time="2020-05-08T16:13:43.591574098Z" level=info msg="Responding (136852 bytes)" func="respond\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:170]" code=200 req.id=bZmE0yvPqGRm req.method=GET req.path=/_matrix/client/r0/sync
time="2020-05-08T16:13:43.669563966Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20190711121626-527ce5ddefc7/json.go:122]" req.id=Gu08hrPPNDr1 req.method=PUT req.path=/_matrix/federation/v1/send/1588819156998
time="2020-05-08T16:13:43.671413876Z" level=info msg="Received transaction \"1588819156998\" containing 0 PDUs, 2 EDUs" func="Send\n\t" file=" [/build/federationapi/routing/send.go:72]" req.id=Gu08hrPPNDr1 req.method=PUT req.path=/_matrix/federation/v1/send/1588819156998
time="2020-05-08T16:13:43.671503032Z" level=warning msg="unhandled edu" func="processEDUs\n\t" file=" [/build/federationapi/routing/send.go:225]" req.id=Gu08hrPPNDr1 req.method=PUT req.path=/_matrix/federation/v1/send/1588819156998 type=m.presence
```

**Comment by Half-Shot** [comment]
as per:

>This does not fix #1015 yet.
